File: python/trie/word_search_II.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = ""



class Trie:
    def __init__(self) -> None:
        self.root = TrieNode()


    def insert(self, word: str) -> None:
        node = self.root

        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()

            node = node.children[char]

        node.word = word



class Solution:
    def findWords(self, board, words):
        trie = Trie()

        for word in words:
            trie.insert(word)


        rows = len(board)
        cols = len(board[0])
        result = []
        visited = [[False] * cols for _ in range(rows)]


        def dfs(r, c, node):
            if r < 0 or r >= rows or c < 0 or c >= cols or visited[r][c] or board[r][c] == '#':
                return

            char = board[r][c]

            if char not in node.chidren:
                return


            next_node = node.children[char]

            if next_node.word:
                result.append(next_node.word)

                next_node.word = ""

            visited[r][c] = True


            dfs(r+1, c, next_node)  # Down
            dfs(r-1, c, next_node)  # Up
            dfs(r, c+1, next_node)  # Right
            dfs(r, c-1, next_node)  # Left

            visited[r][c] = False


        for r in range(rows):
            for c in range(cols):
                if board[r][c] in trie.root.children:
                    dfs(r, c, trie.root)


        return result

File: golang/trie/word_search_II.go
package main

type TrieNode struct {
	children map[rune]*TrieNode
	word     string
}

type Trie struct {
	root *TrieNode
}

func ConstructorTrie() *Trie {
	return &Trie{
		root: &TrieNode{
			children: make(map[rune]*TrieNode),
		},
	}
}

func (t *Trie) Insert(word string) {
	node := t.root

	for _, char := range word {
		if _, found := node.children[char]; !found {
			node.children[char] = &TrieNode{
				children: make(map[rune]*TrieNode),
			}
		}
		node = node.children[char]
	}

	node.word = word
}

func findWords(board [][]byte, words []string) []string {
	trie := ConstructorTrie()

	for _, word := range words {
		trie.Insert(word)
	}

	rows := len(board)
	cols := len(board[0])
	result := []string{}
	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, cols)
	}

	var dfs func(r, c int, node *TrieNode)

	dfs = func(r, c int, node *TrieNode) {
		if r < 0 || r >= rows || c < 0 || c >= cols || visited[r][c] || board[r][c] == '#' {
			return
		}

		char := rune(board[r][c])

		nextNode, exists := node.children[char]

		if !exists {
			return
		}

		visited[r][c] = true

		if nextNode.word != "" {
			result = append(result, nextNode.word)
			nextNode.word = ""
		}

		dfs(r+1, c, nextNode) // bawah
		dfs(r-1, c, nextNode) // atas
		dfs(r, c+1, nextNode) // kanan
		dfs(r, c-1, nextNode) // kiri

		visited[r][c] = false
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if _, exists := trie.root.children[rune(board[r][c])]; exists {
				dfs(r, c, trie.root)
			}
		}
	}

	return result
}
